### ISSUE TITLE
Tapo Smart Hub (H100) connected but no visibility of devices paired with it

### ISSUE THREAD

**Issue by monkga1** [petretiandrea/home-assistant-tapo-p100#354]
Add support for Tapo Smart Hub connected devices such as motion sensors (T100) which don't have own IP addresses.

I'd like to use Tapo Motion Sensors like any other HA supported motion sensors.

**Comment by Piksanhome** [comment]
Support

**Comment by seba1337** [comment]
I have captured some packets, which seems like json with base64 encoded binary data but I don't know how to decode it. Any tips? I can help with reverse engineering as I have H100 with various sensors on it.

**Comment by seba1337** [comment]
OK probably what I've catched was only get status requests from home assistant, meh. I see that everything is encrypted. Seems like a PIA to reverse engineer.

**Comment by seba1337** [comment]
On a RTL-SDR in EU:
![tapo-eu](https://user-images.githubusercontent.com/37665002/218108332-0f7cd74f-b889-4432-8228-70e2fb9cecbd.png)

**Comment by MohitDeshwal** [comment]
Also have hub with few sensors connected willing to help if I could. @seba1337 Do you also have a **Zigbee gateway** around? tried connecting these sensors to that, instead of H100? I am worried about the future of these devices and potential **Matter support** in the Future.

**Comment by seba1337** [comment]
I don't think it's ZigBee compatible.

Anyway I can't find any keyword that I could send to the H100 hub that would get me the connected devices or anything like that. The only thing that works is 

> method: get_device_info

**Comment by petretiandrea** [comment]
Unfortunately without having the hub phisically, I can't develop a proper integration

**Comment by seba1337** [comment]
I have it. How would you approach the problem?

**Comment by Jacxel** [comment]
> Unfortunately without having the hub phisically, I can't develop a proper integration

I am willing to donate the cost of one if you think there is a chance you might be able to get it working. The hub and its accessories work very well and are great value, I would love to be able to integrate them into my HA setup.

**Comment by KennethS2022** [comment]
I'd happily chip in financially as well. Would love to have integration for S220 switches.

**Comment by akshaybengani** [comment]
I am willing to donate, if someone can start working on the integration of Tapo H100, Tapo T315, Tapo S200B

**Comment by Jacxel** [comment]
I used your [plugp100](https://github.com/petretiandrea/plugp100) code to run the `state.get_unmapped_state()` method on the hub and it returned the following:

```
{
   "device_id":"802DD58004D46C425A596926B180734E20B87756",
   "fw_ver":"1.2.3 Build 221012 Rel.103821",
   "hw_ver":"1.0",
   "type":"SMART.TAPOHUB",
   "model":"H100",
   "mac":"29-AE-5A-BB-21-0D",
   "hw_id":"F9GG7B5B3B67E8F51EBF6E2BBDE2977B",
   "fw_id":"00000000000000000000000000000000",
   "oem_id":"941591747A8D77AA74916014FEC3AD90",
   "overheated":false,
   "ip":"111.111.1.1",
   "time_diff":0,
   "ssid":"RsGjw7kvbk1lc2g=",
   "rssi":-58,
   "signal_level":2,
   "in_alarm":false,
   "in_alarm_source":"",
   "latitude":000000,
   "longitude":00000,
   "lang":"en_US",
   "avatar":"hub",
   "region":"Europe/Somwehere",
   "specs":"UK",
   "nickname":"VGFwyGDdWI=",
   "has_set_location_info":true
}
```

I assume it should be possible to create new methods to control the hub and subdevice functionality but I don't really know how, I tried adding a method to set the "in_alarm" parameter to true which i expected to activate the ringer but that didnt seem to happen. 

Note: I have replaced most of the potentially identifying details

**Comment by NJBILLT** [comment]
Any movement on this request?  Is it possible?  I'd love to see the H100 hub integrated with HA.

**Comment by akshaybengani** [comment]
![Screenshot_20230605-071752](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/22613992/f94b96a3-f1ac-4bdd-ba5c-81456048640b)
![Screenshot_20230605-071711](https://github.com/petretiandrea/home-assistant-tapo-p100/assets/22613992/6e27424d-134d-4fc3-a9fe-dcb9100202cd)

Well I think the Tapo team is working on the alexa integration directly, Yesterday I saw a device in my alexa devices section as tapo button, it gives 4 events, tap, double tap, rotate clockwise and rotate counter clockwise. 
I made some alexa routines and it worked for several hours, now again it's not working either some network issue or maybe they are still working on this., So probably we might not need H100 integration directly.

**Comment by monkga1** [comment]
I've always had the H100 attached devices in Alexa. Both of my motion sensors are there and I've been using them in Alexa routines since I got them.  However still no way of having them in HA :(

Regards
Gav
________________________________

**Comment by peculiarvale** [comment]
I'm also willing to donate to have HA integration for H100, to control sensors like T110

**Comment by petretiandrea** [comment]
I will add support for hub, I'm working on it

**Comment by Jacxel** [comment]
> I will add support for hub, I'm working on it

I have a hub and some accessories, I have some basic python knowledge so if I can help I will

**Comment by compton2012** [comment]
If it helps, this repo could probably be easily unpicked to give the required functionality:

https://github.com/mihai-dinculescu/tapo

It breaks down all the api calls neccesary for hub comms and works on the same principle of ip/user/password

**Comment by Mandrake1967** [comment]
+1 for development of H100 support and peripherals

**Comment by NJBILLT** [comment]
> How do you do that?

After adding the TP-Link Tapo integration, the only device that shows is the H100 hub with controls (siren) and diagnostics (overheat and signal level)  Somebody will figure this out to show the devices linked to the hub, but it's way above my pay grade.  Would be nice though as the sensors are dirt cheap and reliable.

**Comment by pbailey008** [comment]
I've been playing around and managed to get a list of child devices that are connected to the hub. I have very limited programming skills but thought I'd share what I found. 

First change was to add get_children to tapo_hub_coordinator
```
from datetime import timedelta

from homeassistant.core import HomeAssistant
from plugp100.api.hub.hub_device import HubDevice
from plugp100.common.functional.either import Either
from plugp100.responses.device_state import DeviceInfo, HubDeviceState
from plugp100.responses.child_device_list import ChildDeviceList

from custom_components.tapo.coordinators import SensorState, TapoCoordinator


class TapoHubCoordinator(TapoCoordinator[HubDeviceState]):
    def __init__(
        self,
        hass: HomeAssistant,
        device: HubDevice,
        polling_interval: timedelta,
    ):
        super().__init__(hass, device, polling_interval)

    def get_sensor_state(self) -> SensorState:
        return SensorState(self.data.info, None, None)

    def get_device_info(self) -> DeviceInfo:
        return self.data.info

    async def _get_state_from_device(self) -> Either[HubDeviceState, Exception]:
        return await self.device.get_state()

    async def get_children(self) -> Either[ChildDeviceList, Exception]:
        return await self._api.get_child_device_list()
```

Then in tapo_hub -> initialize_hub I logged the result of self.hub.get_children():
Results have been scrubbed to remove any unique identifiers...
```
ChildDeviceList(child_device_list=[
{'parent_device_id': 'scrubbed', 'hw_ver': '1.0', 'fw_ver': '1.5.0 Build 230105 Rel.180832', 'device_id': 'scrubbed', 'mac': 'scrubbed', 'type': 'SMART.TAPOSENSOR', 'model': 'T310', 'hw_id': 'scrubbed', 'oem_id': 'scrubbed', 'specs': 'US', 'category': 'subg.trigger.temp-hmdt-sensor', 'bind_count': 1, 'status_follow_edge': False, 'status': 'online', 'lastOnboardingTimestamp': 1691133456, 'rssi': -71, 'signal_level': 3, 'jamming_rssi': -117, 'jamming_signal_level': 1, 'at_low_battery': False, 'temp_unit': 'celsius', 'current_temp': 12.9, 'current_humidity': 78, 'current_temp_exception': 0, 'current_humidity_exception': 18, 'nickname': 'scrubbed', 'avatar': 'sensor_t310', 'report_interval': 16, 'region': 'Australia/Adelaide'}, 
{'parent_device_id': 'scrubbed', 'hw_ver': '1.0', 'fw_ver': '1.10.0 Build 220728 Rel.155144', 'device_id': 'scrubbed', 'mac': 'scrubbed', 'type': 'SMART.TAPOSENSOR', 'model': 'S200B', 'hw_id': 'scrubbed', 'oem_id': 'scrubbed', 'specs': 'US', 'category': 'subg.trigger.button', 'bind_count': 1, 'status_follow_edge': False, 'status': 'online', 'lastOnboardingTimestamp': 1691132148, 'rssi': -72, 'signal_level': 3, 'jamming_rssi': -113, 'jamming_signal_level': 1, 'at_low_battery': False, 'nickname': 'scrubbed', 'avatar': '', 'report_interval': 16, 'region': 'Australia/Adelaide'}, 
{'parent_device_id': 'scrubbed', 'hw_ver': '1.0', 'fw_ver': '1.10.0 Build 220728 Rel.155144', 'device_id': 'scrubbed', 'mac': 'scrubbed', 'type': 'SMART.TAPOSENSOR', 'model': 'S200B', 'hw_id': 'scrubbed', 'oem_id': 'scrubbed', 'specs': 'US', 'category': 'subg.trigger.button', 'bind_count': 1, 'status_follow_edge': False, 'status': 'online', 'lastOnboardingTimestamp': 1691133732, 'rssi': -66, 'signal_level': 3, 'jamming_rssi': -115, 'jamming_signal_level': 1, 'at_low_battery': False, 'nickname': 'scrubbed', 'avatar': 'button', 'report_interval': 16, 'region': 'Australia/Adelaide'}, 
{'parent_device_id': 'scrubbed', 'hw_ver': '1.0', 'fw_ver': '1.8.0 Build 220728 Rel.160245', 'device_id': 'scrubbed', 'mac': 'scrubbed', 'type': 'SMART.TAPOSENSOR', 'model': 'T110', 'hw_id': 'scrubbed', 'oem_id': 'scrubbed', 'specs': 'US', 'category': 'subg.trigger.contact-sensor', 'bind_count': 1, 'status_follow_edge': False, 'status': 'online', 'lastOnboardingTimestamp': 1691133771, 'rssi': -66, 'signal_level': 3, 'jamming_rssi': -115, 'jamming_signal_level': 1, 'at_low_battery': False, 'open': False, 'nickname': 'scrubbed', 'avatar': 'sensor_t110', 'report_interval': 16, 'region': 'Australia/Adelaide'}, 
{'parent_device_id': 'scrubbed', 'hw_ver': '1.0', 'fw_ver': '1.8.0 Build 220728 Rel.160245', 'device_id': 'scrubbed', 'mac': 'scrubbed', 'type': 'SMART.TAPOSENSOR', 'model': 'T110', 'hw_id': 'scrubbed', 'oem_id': 'scrubbed', 'specs': 'US', 'category': 'subg.trigger.contact-sensor', 'bind_count': 1, 'status_follow_edge': False, 'status': 'online', 'lastOnboardingTimestamp': 1691133891, 'rssi': -87, 'signal_level': 1, 'jamming_rssi': -111, 'jamming_signal_level': 1, 'at_low_battery': False, 'open': False, 'nickname': 'scrubbed', 'avatar': 'sensor_t110', 'report_interval': 16, 'region': 'Australia/Adelaide'}
], start_index=0, sum=5)
```

Hope this makes sense, and helps.

**Comment by petretiandrea** [comment]
Thanks for support I have already added support to children in my plugp100 library. The last step is to integrate to homeassistant. I'm working on it

**Comment by CaptainMidnight** [comment]
It would be great if this could give access to the T310 temperature & humidity sensor data, which hang off the hub ......

**Comment by petretiandrea** [comment]
I'm working on It ;)

**Comment by perdimia** [comment]
Great that you are working on it, we bought the hub, doorbell and motion sensors thinking they would work straight away in HA, as we have light strips and a camera that are fully working.
Kinda waste of money right now, fingers crossed you can get this working.

**Comment by RaidentaylorFilipeLeite85** [comment]
I was able to connect the HUB but I cannot access the T110 that is connected to it. 

Is it supposed to work or is it still in development?

**Comment by NJBILLT** [comment]
> Is it supposed to work or is it still in development?

Still in development.  I also have the hub and one door sensor and can only ring the hub's chime; can't see the sensor.  I'm glad I didn't make big investment in these because we may never see it fully implemented.  I'm sure it's a lot of work and there isn't a big demand.  They work fine with the cloud and are cheap.  In lieu of these, I'm using Z-Wave and they work flawlessly with HA.

**Comment by BaerHub** [comment]
Same here, bought the H100 and T100 and thought it would work for HA, hoping for development

**Comment by petretiandrea** [comment]
I'm working on it, but require time. While for sensors an integration is working in progress, for button it's technically complex and infeasible. Let me explain, the integration can poll sensors to get proper data and that's ok, but for button the approach is inverse. When you click a button, it's the button which send a "message" to hub and the hub execute a cloud automation. Actually the hub doesn't expose a local "stream" of events, e.g. button pressed, double press, etcc..., so it's not possible to get event in real time and integration to home assistant. This is the same reason why if the hub is offline the automation does not work, cause rely on cloud instead of locally. An improper solution would be to poll high frequency to the hub to capture events, but this generates a lot of unnecessary network traffic

**Comment by BaerHub** [comment]
> I'm working on it, but require time. While for sensors an integration is working in progress, for button it's technically complex and infeasible. Let me explain, the integration can poll sensors to get proper data and that's ok, but for button the approach is inverse. When you click a button, it's the button which send a "message" to hub and the hub execute a cloud automation. Actually the hub doesn't expose a local "stream" of events, e.g. button pressed, double press, etcc..., so it's not possible to get event in real time and integration to home assistant. This is the same reason why if the hub is offline the automation does not work, cause rely on cloud instead of locally. An improper solution would be to poll high frequency to the hub to capture events, but this generates a lot of unnecessary network traffic

thanks for the explanation!
it would be already amazing, if sensors would work, otherwise my motion sensor is lies around

**Comment by TheTomahawk** [comment]
Well the motion sensor work for me...

``` xml
metadata:
  name: Turn on Kitchen Lights on entering the kitchen .
  description: Turn on Kitchen Lights on entering the kitchen

automations:
  starters:
    - device: Kitchen Sensor - Kitchen
      type: device.state.MotionDetection
      state: motionDetectionEventInProgress
      is: true

  condition:
    type: time.between
    before: sunrise+30min
    after: sunset-60min 
    weekdays:
      - MON
      - TUE
      - WED
      - THU
      - FRI
      - SAT
      - SUN
 
  actions:
    - type: device.command.OnOff
      on: true
      devices: Kitchen Light - Kitchen
```

**Comment by BaerHub** [comment]
i only got 3 entities from H100 but not from the sensor T100 that is connected to the hubAm 06.09.2023 07:20 schrieb TheTomahawk ***@***.***>:
Well the motion sensor work for me...
metadata:
  name: Turn on Kitchen Lights on entering the kitchen .
  description: Turn on Kitchen Lights on entering the kitchen

automations:
  starters:
    - device: Kitchen Sensor - Kitchen
      type: device.state.MotionDetection
      state: motionDetectionEventInProgress
      is: true

  condition:
    type: time.between
    before: sunrise+30min
    after: sunset-60min 
    weekdays:
      - MON
      - TUE
      - WED
      - THU
      - FRI
      - SAT
      - SUN
 
  actions:
    - type: device.command.OnOff
      on: true
      devices: Kitchen Light - Kitchen

**Comment by jamesonuk** [comment]
> I'm working on it, but require time. While for sensors an integration is working in progress, for button it's technically complex and infeasible. Let me explain, the integration can poll sensors to get proper data and that's ok, but for button the approach is inverse. When you click a button, it's the button which send a "message" to hub and the hub execute a cloud automation. Actually the hub doesn't expose a local "stream" of events, e.g. button pressed, double press, etcc..., so it's not possible to get event in real time and integration to home assistant. This is the same reason why if the hub is offline the automation does not work, cause rely on cloud instead of locally. An improper solution would be to poll high frequency to the hub to capture events, but this generates a lot of unnecessary network traffic

Yep I get events are a pain.   I know it is Rust and I haven't actually looked at it but wonder if https://github.com/mihai-dinculescu/tapo/blob/main/src/api/hub_handler.rs might help at all?   Says it took inspiration from your work so could have a similar logical base at some level?

**Comment by petretiandrea** [comment]
> > I'm working on it, but require time. While for sensors an integration is working in progress, for button it's technically complex and infeasible. Let me explain, the integration can poll sensors to get proper data and that's ok, but for button the approach is inverse. When you click a button, it's the button which send a "message" to hub and the hub execute a cloud automation. Actually the hub doesn't expose a local "stream" of events, e.g. button pressed, double press, etcc..., so it's not possible to get event in real time and integration to home assistant. This is the same reason why if the hub is offline the automation does not work, cause rely on cloud instead of locally. An improper solution would be to poll high frequency to the hub to capture events, but this generates a lot of unnecessary network traffic
> 
> Yep I get events are a pain. I know it is Rust and I haven't actually looked at it but wonder if https://github.com/mihai-dinculescu/tapo/blob/main/src/api/hub_handler.rs might help at all? Says it took inspiration from your work so could have a similar logical base at some level?

My implementation is similar, but tapo doesn't offer a stream of events for button "clicks". The only way is polling, but I think is not a good way cause generate a lot of cpu and network waste

**Comment by jamesonuk** [comment]
Ah yeah just dug a little into the code and I can see it grabs trigger logs
https://github.com/mihai-dinculescu/tapo/blob/main/src/api/child_devices/s200b_handler.rs

Can't see that working for a button.   Shame.

**Comment by petretiandrea** [comment]
<img width="742" alt="Immagine 2023-09-12 224447" src="https://github.com/petretiandrea/home-assistant-tapo-p100/assets/25798293/38070c18-062c-401e-81a3-9872c3bef058">

Working in progress 💪

**Comment by petretiandrea** [comment]
> Ah yeah just dug a little into the code and I can see it grabs trigger logs https://github.com/mihai-dinculescu/tapo/blob/main/src/api/child_devices/s200b_handler.rs
> 
> Can't see that working for a button. Shame but totally understand.

Yes, but you get event logs when you ask for them by calling the proper method. An event stream is a push approach to be notified when something change, like button pressure

**Comment by Hatzefatz** [comment]
Hi!

That looks very great!

I'm using those devices already long time just for tapo automation. I did started using Home assistant and can't wait for that integration!


Can you let us know, how long you thing it will take to give us a alpha or beta version for those sensors TP310??


Thank you  very much,
Henrik

**Comment by monisha15081996** [comment]
Hi,
I have managed to pair and integrate the Tapo Hub H100 with Home assistant. The siren works fine. However I am not able to add the Tapo T100 motion sensors since they do not have dedicated ip addresses. Could you please help me integrate the Tapo T100 motion sensors with home assistant?

**Comment by vishn9893** [comment]
> > Ah yeah just dug a little into the code and I can see it grabs trigger logs https://github.com/mihai-dinculescu/tapo/blob/main/src/api/child_devices/s200b_handler.rs
> > Can't see that working for a button. Shame but totally understand.
> 
> Yes, but you get event logs when you ask for them by calling the proper method. An event stream is a push approach to be notified when something change, like button pressure

Ciao Andrea,
Would it be possible to integrate just a portion of the child-devices for the hub H100 as a custom plugin, In my case as well I have a hub h100 and a couple of T100 sensors i guess for the other people above who mentioned as well who have a similar setup needed for detection/monitoring in this case the trigger (event triggers) could be done using the api call that makes an HTTP GET request to the Tapo API endpoint correct?

**Comment by petretiandrea** [comment]
Release 2.9.0 is out now! It supports hub's devices, actually I've tested physically with T31 and S200, but I've extended support to other devices like smart door and motion sensor. Feel free to open new issues

**Comment by DimiTerz** [comment]
Hi, 
thanks for the big update regarding the HUB. I have S210 and S220 light switches paired to the Hub. Atfter 2.9.0 all switches are listed as devices but do't have any controls. Are there plans to expand the Hub support to this devices as well?

**Comment by ondrejsmola** [comment]
Hi,
thanks for your work. I can finally see temp/hum. sensors and I see there exist door sensors. But I don't see state of door sensors. When I look to log, I see some errors. I run HA in in Docker.

`2023-09-21 10:17:18 2023-09-21 10:17:18.312 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo
2023-09-21 10:17:18 Traceback (most recent call last):
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities
2023-09-21 10:17:18     await asyncio.gather(*tasks)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity
2023-09-21 10:17:18     if entity.unique_id is not None:
2023-09-21 10:17:18        ^^^^^^^^^^^^^^^^
2023-09-21 10:17:18   File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id
2023-09-21 10:17:18     return super().unique_id + "_" + self._attr_name.replace(" ", "_")
2023-09-21 10:17:18                                      ^^^^^^^^^^^^^^^^^^^^^^^
2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace'
2023-09-21 10:17:18 2023-09-21 10:17:18.333 ERROR (MainThread) [homeassistant.components.binary_sensor] Error while setting up tapo platform for binary_sensor
2023-09-21 10:17:18 Traceback (most recent call last):
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 370, in _async_setup_platform
2023-09-21 10:17:18     await asyncio.gather(*pending)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities
2023-09-21 10:17:18     await asyncio.gather(*tasks)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity
2023-09-21 10:17:18     if entity.unique_id is not None:
2023-09-21 10:17:18        ^^^^^^^^^^^^^^^^
2023-09-21 10:17:18   File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id
2023-09-21 10:17:18     return super().unique_id + "_" + self._attr_name.replace(" ", "_")
2023-09-21 10:17:18                                      ^^^^^^^^^^^^^^^^^^^^^^^
2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace'
2023-09-21 10:17:18 2023-09-21 10:17:18.366 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Error communication with API: Server disconnected
2023-09-21 10:17:18 2023-09-21 10:17:18.376 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Error communication with API: Server disconnected
2023-09-21 10:17:18 2023-09-21 10:17:18.840 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo
2023-09-21 10:17:18 Traceback (most recent call last):
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities
2023-09-21 10:17:18     await asyncio.gather(*tasks)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity
2023-09-21 10:17:18     if entity.unique_id is not None:
2023-09-21 10:17:18        ^^^^^^^^^^^^^^^^
2023-09-21 10:17:18   File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id
2023-09-21 10:17:18     return super().unique_id + "_" + self._attr_name.replace(" ", "_")
2023-09-21 10:17:18                                      ^^^^^^^^^^^^^^^^^^^^^^^
2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace'
2023-09-21 10:17:18 2023-09-21 10:17:18.955 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo
2023-09-21 10:17:18 Traceback (most recent call last):
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities
2023-09-21 10:17:18     await asyncio.gather(*tasks)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity
2023-09-21 10:17:18     if entity.unique_id is not None:
2023-09-21 10:17:18        ^^^^^^^^^^^^^^^^
2023-09-21 10:17:18   File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id
2023-09-21 10:17:18     return super().unique_id + "_" + self._attr_name.replace(" ", "_")
2023-09-21 10:17:18                                      ^^^^^^^^^^^^^^^^^^^^^^^
2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace'
2023-09-21 10:17:18 2023-09-21 10:17:18.970 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo
2023-09-21 10:17:18 Traceback (most recent call last):
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities
2023-09-21 10:17:18     await asyncio.gather(*tasks)
2023-09-21 10:17:18   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity
2023-09-21 10:17:18     if entity.unique_id is not None:
2023-09-21 10:17:18        ^^^^^^^^^^^^^^^^
2023-09-21 10:17:18   File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id
2023-09-21 10:17:18     return super().unique_id + "_" + self._attr_name.replace(" ", "_")
2023-09-21 10:17:18                                      ^^^^^^^^^^^^^^^^^^^^^^^
2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace'
2023-09-21 10:27:48 2023-09-21 10:27:48.505 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Unexpected exception: Clock moved backwards. Refusing to generate ID.`

**Comment by CaptainMidnight** [comment]
@petretiandrea a massive thank you to you, my temperature and humidity sensors are working perfectly.

Thank you my friend for all your hard work!

Enjoy the coffees

**Comment by NJBILLT** [comment]
Thank you for working on this.,  I see it's working for many.  I only have a contact closure on a door.  With the update, I see it as a device, but the only 2 entities it has are the battery and report interval.  I don't know when it opens and closes.  Am I doing something wrong?  Again a big thank you and I'm sure this will only get better.

**Comment by DominionZA** [comment]
Same with me.

**Comment by RaidentaylorFilipeLeite85** [comment]
> Thank you for working on this., I see it's working for many. I only have a contact closure on a door. With the update, I see it as a device, but the only 2 entities it has are the battery and report interval. I don't know when it opens and closes. Am I doing something wrong? Again a big thank you and I'm sure this will only get better.

Same with me. The state of device ia missing.

**Comment by petretiandrea** [comment]
> S210 and S220

Actually are not supported, please open a new issue where you request support these two devices

**Comment by petretiandrea** [comment]
> Hi, thanks for your work. I can finally see temp/hum. sensors and I see there exist door sensors. But I don't see state of door sensors. When I look to log, I see some errors. I run HA in in Docker.
> 
> `2023-09-21 10:17:18 2023-09-21 10:17:18.312 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo 2023-09-21 10:17:18 Traceback (most recent call last): 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities 2023-09-21 10:17:18 await asyncio.gather(*tasks) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity 2023-09-21 10:17:18 if entity.unique_id is not None: 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id 2023-09-21 10:17:18 return super().unique_id + "_" + self._attr_name.replace(" ", "_") 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace' 2023-09-21 10:17:18 2023-09-21 10:17:18.333 ERROR (MainThread) [homeassistant.components.binary_sensor] Error while setting up tapo platform for binary_sensor 2023-09-21 10:17:18 Traceback (most recent call last): 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 370, in _async_setup_platform 2023-09-21 10:17:18 await asyncio.gather(*pending) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities 2023-09-21 10:17:18 await asyncio.gather(*tasks) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity 2023-09-21 10:17:18 if entity.unique_id is not None: 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id 2023-09-21 10:17:18 return super().unique_id + "_" + self._attr_name.replace(" ", "_") 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace' 2023-09-21 10:17:18 2023-09-21 10:17:18.366 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Error communication with API: Server disconnected 2023-09-21 10:17:18 2023-09-21 10:17:18.376 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Error communication with API: Server disconnected 2023-09-21 10:17:18 2023-09-21 10:17:18.840 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo 2023-09-21 10:17:18 Traceback (most recent call last): 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities 2023-09-21 10:17:18 await asyncio.gather(*tasks) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity 2023-09-21 10:17:18 if entity.unique_id is not None: 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id 2023-09-21 10:17:18 return super().unique_id + "_" + self._attr_name.replace(" ", "_") 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace' 2023-09-21 10:17:18 2023-09-21 10:17:18.955 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo 2023-09-21 10:17:18 Traceback (most recent call last): 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities 2023-09-21 10:17:18 await asyncio.gather(*tasks) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity 2023-09-21 10:17:18 if entity.unique_id is not None: 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id 2023-09-21 10:17:18 return super().unique_id + "_" + self._attr_name.replace(" ", "_") 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace' 2023-09-21 10:17:18 2023-09-21 10:17:18.970 ERROR (MainThread) [homeassistant.components.binary_sensor] Error adding entities for domain binary_sensor with platform tapo 2023-09-21 10:17:18 Traceback (most recent call last): 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 510, in async_add_entities 2023-09-21 10:17:18 await asyncio.gather(*tasks) 2023-09-21 10:17:18 File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 590, in _async_add_entity 2023-09-21 10:17:18 if entity.unique_id is not None: 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 File "/config/custom_components/tapo/hub/binary_sensor.py", line 38, in unique_id 2023-09-21 10:17:18 return super().unique_id + "_" + self._attr_name.replace(" ", "_") 2023-09-21 10:17:18 ^^^^^^^^^^^^^^^^^^^^^^^ 2023-09-21 10:17:18 AttributeError: 'NoneType' object has no attribute 'replace' 2023-09-21 10:27:48 2023-09-21 10:27:48.505 ERROR (MainThread) [custom_components.tapo.coordinators] Error fetching tapo data: Unexpected exception: Clock moved backwards. Refusing to generate ID.`

Please open a specific issue

**Comment by petretiandrea** [comment]
I will close this issue, cause some devices are actually integrated with HA. Each specific issue related to a HUB child device must be reported to a new issue

> Thank you for working on this., I see it's working for many. I only have a contact closure on a door. With the update, I see it as a device, but the only 2 entities it has are the battery and report interval. I don't know when it opens and closes. Am I doing something wrong? Again a big thank you and I'm sure this will only get better.



> Thank you for working on this., I see it's working for many. I only have a contact closure on a door. With the update, I see it as a device, but the only 2 entities it has are the battery and report interval. I don't know when it opens and closes. Am I doing something wrong? Again a big thank you and I'm sure this will only get better.

Refers to #530

**Comment by petretiandrea** [comment]
I will close this issue cause child devices are actually supported. Any issue related to specific child device require a new issue

**Comment by bouledrouasam** [comment]
is there any solution for the smart button s200b

**Comment by ados8** [comment]
For those looking for any workaround I found a way.
https://github.com/petretiandrea/home-assistant-tapo-p100/pull/784#issuecomment-2480470653

**Comment by erosen14** [comment]
PR open for this: https://github.com/petretiandrea/home-assistant-tapo-p100/pull/934

Adds EventEntity support for S200B button presses and S200D button + dial rotation. Creates automations-ready entities that fire single_click and rotation (with degrees attribute) events.

The main concern raised in this thread is polling overhead. The implementation addresses this with a budget-based adaptive polling algorithm — instead of a fixed polling interval, it measures the actual hub API latency and jitter per cycle, then dynamically computes the interval as (latency + 2×jitter) / utilization_target. This keeps hub load within a configurable budget (default 35%, adjustable via a slider entity). In practice this means ~100ms polling when the hub is idle, backing off automatically under load.

Importantly, the hub's event log endpoint is what gets polled — not the buttons themselves. The buttons communicate with the hub over sub-GHz radio and are unaffected by polling frequency, so there's no battery impact.

Not a perfect solution (polling will never match a push-based protocol), but in practice button response is <200ms and dial control is smooth enough for light dimming.

**Example blueprints** (one-click import into HA): https://gist.github.com/erosen14/8e427e9da6351abf8554cfc12bea421d
